static/js/subs.js

```diff
@@ -707,31 +730,67 @@ exports.close = function () {
-    var switch_row; // initialize var for row that we're switching to
-    // if rows have undefined attributes (no selected/active row)
-    if (!active_row.attr('data-stream-id')) {
-        switch_row = $('div.stream-row:first-child'); // set active row to first row
+exports.switch_rows = function (event) {
+    var active_data = get_active_data();
+    var switch_row;
+    if (!active_data[1] || active_data[0].hasClass('notdisplayed')) {
+        switch_row = $('div.stream-row:not(.notdisplayed):first');
     } else if (event === 'up_arrow') {
-        switch_row = active_row.prev(); // previous row
+        switch_row = active_data[0].prev();
     } else if (event === 'down_arrow') {
-        switch_row = active_row.next(); // next row
+        switch_row = active_data[0].next();
+        if ($('#search_stream_name').is(":focus")) {
+            $('#search_stream_name').blur();
+            return;
+        }
     }
 
-    var switch_row_id = switch_row.attr('data-stream-id');
-    // if both ID and row are defined, making sure to escape hidden rows
-    if (switch_row_id && !switch_row.hasClass('notdisplayed')) {
-        var switch_row_name = stream_data.get_sub_by_id(switch_row_id).name;
-        var hash = ['#streams', switch_row_id, switch_row_name]; // set hash
-        var hash_components = { // hash_components to send to subs.change_state()
-            base: hash.shift(),
-            arguments: hash,
-        };
-        exports.change_state(hash_components);
+    var row_data = get_row_data(switch_row);
+    if (row_data && !switch_row.hasClass('notdisplayed')) {
+        var switch_row_name = row_data[1].name;
+        var hash = ['#streams', row_data[0], switch_row_name];
+        export_hash(hash);
+    } else if (event === 'up_arrow' && !row_data) {
+        $('#search_stream_name').focus();
     }
 };
 
+exports.keyboard_sub = function () {
+    var active_data = get_active_data();
+    var row_data = get_row_data(active_data[0]);
+    if (row_data) {
+        subs.sub_or_unsub(row_data[1]);
+        if (row_data[1].subscribed && active_data[2].text() === 'Subscribed') {
+            active_data[0].addClass('notdisplayed');
+            active_data[0].removeClass('active');
+        }
+    }
+};
+
+exports.toggle_view = function (event) {
+    var active_data = get_active_data();
+    var hash;
+    if (event === 'right_arrow' && active_data[2].text() === 'Subscribed') {
+        hash = ['#streams', 'all'];
+        export_hash(hash);
+    } else if (event === 'left_arrow' && active_data[2].text() === 'All streams') {
+        hash = ['#streams', 'subscribed'];
+        export_hash(hash);
+    }
+};
+
+exports.view_stream = function () {
+    var active_data = get_active_data();
+    var row_data = get_row_data(active_data[0]);
+    if (row_data) {
+        window.location.hash = '#narrow/stream/' + row_data[1].name;
+    }
+};
+
+exports.new_stream_clicked = function () {
+    $("#create_stream_button").click();
+};
```

Comment:
It might be cleaner to refactor what is currently the click handler into a function that we can also here, rather than doing the click indirection.